File: Graph/d.EvaluateDivision.cpp
class Solution {
public:
    map<string , string> parent;
    map<string , double> multiples;
    void addToDSU(string a){
        parent[a] = a;
        multiples[a] = 1.0;
    }
    string find(string x){
        if(parent[x] == x){
            return parent[x];
        }else{
            string currentParent = parent[x];
            string finalParent = find(currentParent);
            parent[x] = finalParent;
            // compression
            multiples[x] = multiples[x] * multiples[currentParent];
            return finalParent;
        }
    }
    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        for(vector<string> equation : equations){
            string operand1 = equation[0];
            string operand2 = equation[1];
            addToDSU(operand1);
            addToDSU(operand2);
        }
        // Union
        int i = 0;
        for(vector<string> equation : equations){
            string l0 = find(equation[0]);
            string l1 = find(equation[1]);
            
            if(l0 != l1){
                double multiple1 = multiples[equation[0]];
                double multiple2 = multiples[equation[1]];
                parent[l0] = l1;
                multiples[l0] = (values[i] * multiple2) / multiple1;
            }
            i++;
        }
        // finding answers
        vector<double> ans;
        for(vector<string> query : queries){
            string operand1 = query[0];
            string operand2 = query[1];
        
            
            // first case - Operand hi na ho
            if(parent.find(operand1) == parent.end() || parent.find(operand2) == parent.end()){ 
                ans.push_back(-1.0);
                continue;
            }
            
            // second case - agar dono ka lead same hi na ho
            string lead1 = find(operand1);
            string lead2 = find(operand2);
            if(lead1 != lead2){
                ans.push_back(-1.0);
                continue;
            }
            // third case
            double multiple1 = multiples[operand1];
            double multiple2 = multiples[operand2];
            double mul = multiple1 / multiple2;
            ans.push_back(mul);
        }
        return ans;
    }
};

File: Graph/d.EvaluateDivision.java
class Solution {
    // DSU
    HashMap<String , String> parent;
    HashMap<String , Double> multiple;   
    void addToDSU(String x) {
        if(parent.containsKey(x) == false){
            parent.put(x , x);
            multiple.put(x , 1.0);
        }
    }
    String find(String x) {
        if(parent.get(x).equals(x)){
            return parent.get(x);
        }else{
            String currentParent = parent.get(x);
            String finalParent = find(currentParent);
            parent.put(x , finalParent);
            ///compression
            multiple.put(x , multiple.get(currentParent) * multiple.get(x));
            return finalParent;
        }
    }
    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
        parent = new HashMap<>();
        multiple = new HashMap<>();
        for(List<String> equation : equations){
            addToDSU(equation.get(0));
            addToDSU(equation.get(1));
        }
        int i = 0;
        for(List<String> equation : equations){
            String l0 = find(equation.get(0));
            String l1 = find(equation.get(1));
            
            if(l0.equals(l1) == false) {
                // union
                double m0 = multiple.get(equation.get(0));
                double m1 = multiple.get(equation.get(1));
                parent.put(l0 , l1);
                multiple.put(l0 , (values[i] * m1)/ m0);
            }
            i++;
        }
        double[] res = new double[queries.size()];
        i = 0;
        for(List<String> query : queries){
            String q0 = query.get(0);
            String q1 = query.get(1);
            
            if(parent.containsKey(q0) == false || parent.containsKey(q1) == false){
                res[i] = -1;
                i++;
                continue;
            }
            
            String l0 = find(q0);
            String l1 = find(q1);
            if(l0.equals(l1) == false){
                res[i] = -1;
                i++;
                continue;
            }
            double m0 = multiple.get(q0);
            double m1 = multiple.get(q1);
            res[i] = m0 / m1;
            i++;
        }
        return res;
    }
}